src/compile/projection/parse.ts

```diff
@@ -0,0 +1,140 @@
+import {SHAPE, X, Y} from '../../channel';
+import {Config} from '../../config';
+import {MAIN} from '../../data';
+import {isFieldDef} from '../../fielddef';
+import {GEOSHAPE} from '../../mark';
+import {Projection, PROJECTION_PROPERTIES} from '../../projection';
+import {LookupData} from '../../transform';
+import {LATITUDE, LONGITUDE} from '../../type';
+import {contains, duplicate, every} from '../../util';
+import {VgProjection, VgSignalRef} from '../../vega.schema';
+import {isUnitModel, Model} from '../model';
+import {UnitModel} from '../unit';
+import {ProjectionComponent} from './component';
+
+export function parseProjection(model: Model) {
+  if (isUnitModel(model)) {
+    model.component.projection = parseUnitProjection(model);
+  } else {
+    // because parse happens from leaves up (unit specs before layer spec),
+    // we can be sure that the above if statement has already occured
+    // and therefore we have access to child.component.projection
+    // for each of model's children
+    model.component.projection = parseNonUnitProjections(model);
+  }
+}
+
+function parseUnitProjection(model: UnitModel): ProjectionComponent {
+  const {specifiedProjection, markDef, config, encoding} = model;
+
+  const isGeoShapeMark = markDef && markDef.type === GEOSHAPE;
+  const isGeoPointMark = encoding && [X, Y].every(
+    (channel) => {
+      const def = encoding[channel];
+      return isFieldDef(def) && contains([LATITUDE, LONGITUDE], def.type);
+  });
+
+  if (isGeoShapeMark || isGeoPointMark) {
+    let data: VgSignalRef | string;
+    if (isGeoPointMark || (isGeoShapeMark && encoding[SHAPE])) {
+      // if there is a shape encoding or latitude / longitude encoding,
+      // the compiler added a geojson transform w/ a signal as specified
+      data = {
+        signal: model.getName('geojson')
+      };
+    } else {
+      // main source is geojson, so we can just use that
+      data = model.requestDataName(MAIN);
+    }
+
+    return new ProjectionComponent(model.projectionName(true), {
+      ...(config.projection || {}),
+      ...(specifiedProjection || {}),
+    } as Projection, [model.getSizeSignalRef('width'), model.getSizeSignalRef('height')], [data]);
```

Comment:
Casting can cloak compile error for type mismatches. 
Thus, we generally avoid casting `as` unless it is really needed.  (e.g., when we run into TS bugs.)